libs/cluster/Session/ClusterCommands.cs

```diff
@@ -58,10 +58,14 @@ private List<byte[]> GetKeysInSlot(int slot, int keyCount)
             return keys;
         }
 
-        private bool ParseSlots(int count, ref byte* ptr, out HashSet<int> slots, out ReadOnlySpan<byte> resp, bool range)
+        /// <summary>
+        /// Try to parse slots
+        /// </summary>
+        /// <param name="errorMessage">The ASCII encoded error message if the method return <c>false</c>; otherwise <c>default</c></param>
+        private bool TryParseSlots(int count, ref byte* ptr, out HashSet<int> slots, out ReadOnlySpan<byte> errorMessage, bool range)
```

Comment:
The state this method holds is extremely complex.. Am I right thinking that it is quietly swallowing errors and is that intended behavior or not?

Reply:
After staring at this for a while, it seems the return value boolean that flows from cluster command processing will trigger the unknown subcommand error. However we should/can exit the loop as soon as we know that there is for example repeat slots.. Got tests failures from that which is nice 👍

Reply:
I think the way I can return original behavior here is by restoring the behavior where:
* If ``TryParseSlots`` returns `false`:
  * There was wrong number of arguments, no errors thrown here but a "wrong number of arguments" will be thrown at higher layer.
* If ``TryParseSlots`` returns `true`:
  * If the out parameter ``errorMessage == default`` evals to ``true``, we have error to write, however don't short-circuit to return false that would also incorrectly write "wrong num of parameters" fallback error. (I had this behavior here in the PR for while)
  * If the out parameter ``errorMessage == default`` evals to ``false``, everything is good, continue.

Reply:
I just saw your messages. When in doubt you can check the Redis protocol for the corresponding command. Of course, some functionality maybe missing but we try our hardest to much the original protocol specification.
I think you nailed down the overall workflow, just summarizing the keys points of ParseSlots below.

- If slot provided is in range else return error.
- I slot provided is specified multiple times else return error. The error here is weird because we are using a set so if a slot is specified multiple times we can still continue without any issues but it a restriction of the Redis protocol.
- We are trying to do everything in place, meaning that we keep reading the slots and gathering possible errors in order to drain the receive buffer. It is a left over from before any attempts to refactor the parser which are still ongoing. Ideally, we should fail immediately, and make the caller is responsible for draining.

I think the last point is what making the ParseSlots method so complicated. If it gets in the way of having a common WriteError method, should be fine to leave it as is until we do a more comprehensive refactoring of the parser.

Reply:
An edge condition that we are not tested is what happens when the input contains a non-numeric value for a slot. But I guess, this is an issue we need to revisit because it is unrelated to the current PR.

Reply:
> * I slot provided is specified multiple times continue processing else capture error. This condition is weird and possibly does not make sense to have since we consolidate the slots into a set. We could omit it but I am worried that because it is part of the specification, somebody will create an issue.

I believe the spec says:
> The command fails if the same slot is specified multiple times.

For both [DELSLOTS](https://redis.io/commands/cluster-delslots/) and [ADDSLOTS](https://redis.io/commands/cluster-addslots/)..  So presumable we can just bailout instantly with ``return false``, and let the command draining handle rest of args?

However the other paths which hit this are the ranges and I'm not clear where to read what we should do for overlapping ranges for example? The slot range variants don't specify failure for overlapping if I'm reading [this](https://redis.io/commands/cluster-addslotsrange/) right?

> * Last condition is that we are processing inputs in place. Therefore, we need to capture the errors and continue processing to drain the recv buffer. Ideally, we should fail immediately, and make the caller is responsible for draining.

I now read this and yea, it'd be ideal if we didn't have to care about draining the arguments here 🤔

Reply:
I'll restore the proper argument draining behavior here.

Reply:
> I'll restore the argument draining behavior here.
Yes, do that. I was going to suggest, that we should probably not change the behavior of ParseSlots since this PR has a different purpose.

Reply:
> > * I slot provided is specified multiple times continue processing else capture error. This condition is weird and possibly does not make sense to have since we consolidate the slots into a set. We could omit it but I am worried that because it is part of the specification, somebody will create an issue.
> 
> I believe the spec says:
> 
> > The command fails if the same slot is specified multiple times.
> 
> For both [DELSLOTS](https://redis.io/commands/cluster-delslots/) and [ADDSLOTS](https://redis.io/commands/cluster-addslots/).. So presumable we can just bailout instantly with `return false`, and let the command draining handle rest of args?
> 
> However the other paths which hit this are the ranges and I'm not clear where to read what we should do for overlapping ranges for example? The slot range variants don't specify failure for overlapping if I'm reading [this](https://redis.io/commands/cluster-addslotsrange/) right?
> 
> > * Last condition is that we are processing inputs in place. Therefore, we need to capture the errors and continue processing to drain the recv buffer. Ideally, we should fail immediately, and make the caller is responsible for draining.
> 

Just to note, this is not a redis requirement more of our way to do things fast in-place without extra allocations.

> I now read this and yea, it'd be ideal if we didn't have to care about draining the arguments here 🤔

We have DrainCommand that we can use, though to do it correctly we need to keep track how many of the args we read correctly before the error, so we can drain the rest.

Maybe a more appropriate solution will be a method that reads all parameters for ADDSLOTS/DELSLOTS in a array/list (ensuring that everything is parsed/drained) and then validate the list input.
It might be fine for perf reasons since CLUSTER commands are in the slow path.